### DISCUSSION TITLE
Fast Refresh support with preserved local state for coffeescript

### DISCUSSION THREAD

**Discussion by Cottin** [vercel/next.js#33810] [Ideas]
### Describe the feature you'd like to request

You can add a `coffee-loader` to the `webpack` part of `next.config.js` and Fast refresh works okey, but local state inside components are lost. It would be nice if the local state could be kept.

I realise this will be so down prioritised, it would never happen, but could you give me a tip for how to implement it myself? See the technical details of the problem below.

### Describe the solution you'd like

Local state inside components being kept in coffeescript file during Fast Refresh.

### Describe alternatives you've considered

Given a basic setup like so:

	config.module.rules.push({
		test: /\.(coffee)$/,
		include: [dir],
		exclude: /node_modules/,
		use: [
			defaultLoaders.babel,
			{
				loader: 'coffee-loader',
				options: {
					bare: true,
				}
			},
		],
	})

coffee-loader will compile to "fancy" js and `defaultLoaders.babel` will compile to "basic" js + **TRY TO** add the fast refresh stuff like

```
var _c, _c1;
$RefreshReg$(_c, "Home");
$RefreshReg$(_c1, "Hello");
```

But it will fail, and it seems it's because it's looking for a pattern of a functional component like so:
```
var Hello = function() {...
```
and coffeescript moves all `var`s to the top of the file while compiling like so:

```
var Hello, OtherComponent;
Hello = function Hello() { ...
```
and thus the pattern is not matched.

If we manually "improve" the output from the coffeescript compiler (before piping it in to the defaultLoaders.babel loader) by adding back the `var` like so :
```
var OtherComponent;
var Hello = function() {...
```
the `$RefreshReg$(_c1, "Hello");` gets added and Fast Refresh works with preserving the local state.

Is there any way to configure the `defaultLoaders.babel` to recognise patterns without the `var` as a component in order to add the  `$RefreshReg$(_c1, "Hello");` ?

**Comment by Cottin** [comment]
If there are any other coffee aficionados who happen to find this page, a simple work around seems to be to create a custom loader for webpack:

```
// fastCoffeeLoader.js
module.exports = function fastCoffeeLoader(source) {
  return source.replace(/^([A-Z]\w+)\s=\sfunction/gm, 'var $1 = function');
}

// next.config.js
module.exports = (phase, { defaultConfig }) => {
	return {
	  reactStrictMode: true,
		pageExtensions: ['coffee', 'jsx', 'js', 'ts', 'tsx'],

		webpack(config, options) {
			const { dir, defaultLoaders } = options
			config.resolve.extensions.push('.coffee')

			config.module.rules.push({
				test: /\.(coffee)$/,
				include: [dir],
				exclude: /node_modules/,
				use: [
					defaultLoaders.babel,
					{loader: path.resolve(__dirname, './fastCoffeeLoader.js')} // <----------
					{loader: 'coffee-loader'},

				]
```

It turns the output of coffee-loader:

```
var Hello, normalFunction;
Hello = function Hello() { ...
normalFunction = function normalFunction() { ...
```

into

```
var Hello, normalFunction;
var Hello = function() {...
normalFunction = function normalFunction() { ...
```

before passing it into the `defaultLoaders.babel` who then does its magic and adds the `var _c1;$RefreshReg$(_c1, "Hello");` stuff and hot refresh starts to work while preserving state in coffee components too.

**Reply by kenOfYugen** [reply]
For this workaround you match the function declaration
```
var Hello, OtherComponent;
Hello = function() { ...
```

and you prefix it with `var `
```
var Hello, OtherComponent;
var Hello = function() { ...
```

Does the `/^([A-Z]\w+)\s=\sfunction/gm` RegEx you have provided match the declaration?
In my case it doesn't, but `/^(\w+)\s=\sfunction/gm` does match it.
Am I understanding this correctly?
